### PR TITLE
Add documentation for OAuth2 generic authenticator

### PR DESCRIPTION
**Pull request by Avarjana**
## Description
Add documentation for OAuth2GenericAuthenticator with Kakao. 

## Related issues: 
- https://github.com/wso2/product-is/issues/14166

## Related PRs: 
- https://github.com/wso2-extensions/identity-outbound-auth-oauth2/pull/16
- https://github.com/wso2/product-is/pull/14234
- https://github.com/wso2/samples-is/pull/338

### CODE REVIEW

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+|[protected void processAuthenticationResponse(HttpServletRequest request,HttpServletResponse response, AuthenticationContext context)](https://github.com/wso2-extensions/identity-outbound-auth-oauth2/blob/75745516bb81f3b93b47075da720cd2fbc194217/components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java#L116-L145)|Implements the logic of the custom federated authenticator.|
+|[protected void buildClaims(AuthenticationContext context, String userInfoString)](https://github.com/wso2-extensions/identity-outbound-auth-oauth2/blob/75745516bb81f3b93b47075da720cd2fbc194217/components/org.wso2.carbon.identity.application.authenticator.oauth2/src/main/java/org.wso2.carbon.identity.application.authenticator.oauth2/Oauth2GenericAuthenticator.java#L147-L183)|Implements the logic for extracting the subject from claims and for other claim mappings.|
+
+---
+
+## Deploy the custom federated authenticator in WSO2 IS
```
Comment (piraveena) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:34]:
```suggestion
## Deploy the custom OAuth2 federated authenticator in WSO2 IS
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+1. Once the implementation is done, navigate to the root of your project and run the following command to compile the
+   service.
+    ``` xml
+    mvn clean install
+    ```
+2. Copy the compiled `jar` file inside `/target` and paste it to the `<IS_HOME>/repository/components/dropins`.
+
+---
+
+## Configure custom federated authenticator
```
Comment (piraveena) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:45]:
```suggestion
## Configure custom OAuth2 federated authenticator
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+1. To configure the custom authenticator, click the Add button under Identity Providers and add an IDP name to register the new IDP.
+2. Click Federated Authenticators and expand `OAuth2 Custom Authenticator Configuration`. Configure it as follows.
```
Comment (piraveena) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:48]:
```suggestion
2. Click Federated Authenticators and expand `Custom OAuth2 Authenticator Configuration`. Configure it as follows.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+    - Select the **Enable** and **Default** check boxes.
+    - **Client Id** - The client ID value generated by the external OAuth2 IdP.
+    - **Client Secret** - The client secret value generated by the external OAuth2 IdP.
+    - **Callback Url** - https://localhost:9443/commonauth
+    - **Authorization Endpoint URL** - https://kauth.kakao.com/oauth/authorize
```
Comment (piraveena) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:54]: do we need to configure them? These needs to go inside the specific authenticator while deploying sicne these are authenticator specific endpoints

Comment (piraveena) [reply en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:54]: Do we have such behavior in Kakao also? Do they need to explicitly configure the endpoint?

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 5]
```diff
@@ -0,0 +1,84 @@
+# Write a Custom OAuth2 Federated Authenticator
+
+A custom OAuth2 authenticator can be written to authenticate a user with an external system. The external system can be
+any Identity provider supporting OAuth2 including Facebook, Kakao, Naver etc. You can use the OAuth2GenericAuthenticator
```
Comment (pavinduLakshan) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:4]: IMO, it would be great if the text "OAuth2GenericAuthenticator" is a hyperlink that points to the class location in the GH repo.

Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:4]:
```suggestion
any Identity provider supporting OAuth2 including Facebook, Kakao, Naver, etc. You can use the `OAuth2GenericAuthenticator`
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+## Steps to write a custom OAuth2 authenticator
+
+1. First create a maven project for the custom OAuth2 authenticator. Refer to
+   the [pom.xml](https://github.com/wso2/samples-is/blob/master/authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml)
+   file used for the Custom authenticator.
+2. Refer
+   the [service component class](https://github.com/wso2/samples-is/blob/master/authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java)
+   as well since the authenticator is written as an OSGI service to deploy in WSO2 Identity Server and register it as an
+   authenticator
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:15]:
```suggestion
1. Create a maven project for the custom OAuth2 authenticator. Refer to the following files:

> - the [pom.xml](https://github.com/wso2/samples-is/blob/master/authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/pom.xml) file used for the Custom authenticator.
> - The [service component class](https://github.com/wso2/samples-is/blob/master/authenticators/components/org.wso2.carbon.identity.sample.oauth2.federated.authenticator/src/main/java/org/wso2/carbon/identity/sample/oauth2/federated/authenticator/internal/OAuth2CustomAuthenticatorServiceComponent.java) as well since the authenticator is written as an OSGI service to deploy in WSO2 Identity Server and register it as an authenticator
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+3. The custom authenticator should be written by extending the OAuth2GenericAuthenticator class.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:16]:
```suggestion
2. The custom authenticator should be written by extending the `OAuth2GenericAuthenticator` class.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+4. You can find a [Kakao custom authenticator](https://github.com/wso2-extensions/identity-outbound-auth-kakao) here for
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:17]:
```suggestion
3. You can find a [Kakao custom authenticator](https://github.com/wso2-extensions/identity-outbound-auth-kakao) here for
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+---
+
+## Deploy the custom OAuth2 federated authenticator in WSO2 IS
+
+1. Once the implementation is done, navigate to the root of your project and run the following command to compile the
+   service.
+    ``` xml
+    mvn clean install
+    ```
+2. Copy the compiled `jar` file inside `/target` and paste it to the `<IS_HOME>/repository/components/dropins`.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:41]:
```suggestion
To deploy the custom OAuth2 federated authenticator:
1. Open a terminal, navigate to the root of your project, and compile the service by running the following command:
    ``` xml
    mvn clean install
    ```
2. Copy the compiled `jar` file inside `/target` folder and paste it to the `<IS_HOME>/repository/components/dropins`.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+---
+
+## Configure custom OAuth2 federated authenticator
+
+1. To configure the custom authenticator, click the Add button under Identity Providers and add an IDP name to register the new IDP.
+2. Click Federated Authenticators and expand `Custom OAuth2 Authenticator Configuration`. Configure it as follows.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:48]:
```suggestion
2. Expand **Federated Authenticators > Custom OAuth2 Authenticator Configuration** and configure it as follows:
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    ``` xml
+    mvn clean install
+    ```
+2. Copy the compiled `jar` file inside `/target` and paste it to the `<IS_HOME>/repository/components/dropins`.
+
+---
+
+## Configure custom OAuth2 federated authenticator
+
+1. To configure the custom authenticator, click the Add button under Identity Providers and add an IDP name to register the new IDP.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:47]:
```suggestion
1. To configure the custom authenticator, initially [register a new identity provider]({{base_path}}/guides/identity-federation/add-idp/).
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+## Configure an application with the custom federated authenticator
+
+1. Start the server and log in to the WSO2 IS Management Console.
+2. Navigate to `Main > Identity > Service Providers > Add`.
+3. Add a Service Provider name and optionally, add a description.
+4. Click `Register`. (Let’s use the playground app and refer this to configure the playground app.
+5. List the Service Providers and edit the service provider as follows by navigating to the `OAuth/OpenID Configuration` section under `Inbound Authentication Configuration` as explained above.
+6. Click Configure and add `http://localhost:8080/playground2/oauth2client` as the call back URL. Click Update.
+7. Navigate to `Local & Outbound Authentication Configuration` and you can find the Authentication Type. Select `Federated Authentication` and select the configured authenticator and update to save the changed configurations.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:74]:
```suggestion
1. On the WSO2 IS Management Console, go to `Main > Identity > Service Providers > Add`.
2. Enter a Service Provider name and click **Register** to add the new service provider.
3. Expand **Inbound Authentication Configuration > OAuth/OpenID Connect Configuration** and click **Configure** 4. Enter `http://localhost:8080/playground2/oauth2client` as the **Callback URI**, and click **Update**.
5. Expand to **Local & Outbound Authentication Configuration** and select the configured authenticator from the **Federated Authentication** list.
6. Click **Update** to save the configurations.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+---
+
+## Try out the scenario
+
+1. Access the playground app by using [http://localhost:8080/playground2](http://localhost:8080/playground2).
+2. This will redirect to the login page of Kakao.
+3. Enter email and password for Kakao.
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:82]:
```suggestion
3. Enter the email address and password for Kakao.

You will be successfully logged into the application.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 14]
```diff
@@ -0,0 +1,84 @@
+# Write a Custom OAuth2 Federated Authenticator
+
+A custom OAuth2 authenticator can be written to authenticate a user with an external system. The external system can be
+any Identity provider supporting OAuth2 including Facebook, Kakao, Naver etc. You can use the OAuth2GenericAuthenticator
+available in WSO2 Identity Server to create custom OAuth2 authenticators.
+
+## Steps to write a custom OAuth2 authenticator
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:7]:
```suggestion
## Write a custom OAuth2 federated authenticator
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+1. Once the implementation is done, navigate to the root of your project and run the following command to compile the
+   service.
+    ``` xml
+    mvn clean install
+    ```
+2. Copy the compiled `jar` file inside `/target` and paste it to the `<IS_HOME>/repository/components/dropins`.
+
+---
+
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:44]:
```suggestion
## Configure the Identity Server

This section guides you on how to configure the identity server to use the custom OAuth2 federated authenticator.
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+## Configure custom OAuth2 federated authenticator
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:45]:
```suggestion
### Configure custom OAuth2 federated authenticator
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+
+
+    !!!info
+         Here, the Client Id and Client Secret are the values of Kakao.
+
+   ![Federated Oauth2 Authenticator]({{base_path}}/assets/img/extend/federated-oauth2-authenticator.png)
+
+---
+
+## Configure an application with the custom federated authenticator
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:66]:
```suggestion
### Configure the SP with the custom federated authenticator
```

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+2. Navigate to `Main > Identity > Service Providers > Add`.
+3. Add a Service Provider name and optionally, add a description.
+4. Click `Register`. (Let’s use the playground app and refer this to configure the playground app.
+5. List the Service Providers and edit the service provider as follows by navigating to the `OAuth/OpenID Configuration` section under `Inbound Authentication Configuration` as explained above.
+6. Click Configure and add `http://localhost:8080/playground2/oauth2client` as the call back URL. Click Update.
+7. Navigate to `Local & Outbound Authentication Configuration` and you can find the Authentication Type. Select `Federated Authentication` and select the configured authenticator and update to save the changed configurations.
+
+---
+
+## Try out the scenario
```
Comment (divyaamunugama) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:78]:
```suggestion
## Try it out
```

**Review by divyaamunugama — APPROVED** [review]

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+
+   ![Federated Oauth2 Authenticator]({{base_path}}/assets/img/extend/federated-oauth2-authenticator.png)
+
+---
+
+### Configure the SP with the custom federated authenticator
+
+1. On the WSO2 IS Management Console, go to **Main > Identity > Service Providers > Add**.
+2. Enter a Service Provider name and click **Register** to add the new service provider.
+3. Expand **Inbound Authentication Configuration > OAuth/OpenID Connect Configuration** and click **Configure**
```
Comment (sachiniWettasinghe) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:69]: 3. Expand **Inbound Authentication Configuration > OAuth/OpenID Connect Configuration** and click **Configure**.

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md [hunk 20]
```diff
@@ -0,0 +1,84 @@
...
+    - **Client Id** - The client ID value generated by the external OAuth2 IdP.
+    - **Client Secret** - The client secret value generated by the external OAuth2 IdP.
+    - **Callback Url** - https://localhost:9443/commonauth
+    - **Authorization Endpoint URL** - https://kauth.kakao.com/oauth/authorize
+    - **Token Endpoint URL** - https://kauth.kakao.com/oauth/token
+    - **User Information Endpoint URL** - https://kapi.kakao.com/v2/user/me
+
+
+    !!!info
+         Here, the Client Id and Client Secret are the values of Kakao.
```
Comment (sachiniWettasinghe) [inline en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:59]: Don't we need to point here the guidelines to take these values from Kakao?

Comment (Avarjana) [reply en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:59]: At a point where writing a custom OAuth2 authenticator, we expect the developer to know the basics of the external OAuth2 IdP. Therefore, we can rephrase this to be more generic with,
`Here, the Client Id and Client Secret are the values generated by the external OAuth2 IdP.`
What do you think?

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md (hunk 19 above)
Comment (Avarjana) [reply en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:69]: Fixed with https://github.com/wso2/docs-is/pull/3421

File: en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md (hunk 20 above)
Comment (Avarjana) [reply en/docs/references/extend/federation/write-a-custom-oauth2-federated-authenticator.md:59]: Changed with https://github.com/wso2/docs-is/pull/3421
